jumanji/environments/routing/macvrp/utils.py

```diff
@@ -0,0 +1,274 @@
+    # The early and late penalties are only calculated if the local time is outside the window.
+    # If the local time is inside the window, the early and late penalties are zero.
+    early_penalty = jax.numpy.where(
+        local_times < window_start,
+        (window_start - local_times) * early_coefs,
+        0,
+    )
+    late_penalty = jax.numpy.where(
+        local_times > window_end,
+        (local_times - window_end) * late_coefs,
+        0,
+    )
+    time_penalties = early_penalty + late_penalty
+
+    return time_penalties
+
+
+def compute_distance(
+    start_node_coordinates: chex.Array, end_node_coordinates: chex.Array
+) -> jax.numpy.float32:
+    """Calculate the distance traveled between two nodes"""
+    return jax.numpy.linalg.norm(
+        (start_node_coordinates - end_node_coordinates), axis=1
+    )
+
+
+def generate_problem(
+    key: chex.PRNGKey,
+    num_customers: jax.numpy.int16,
+    total_capacity: jax.numpy.int16,
+    map_max: jax.numpy.float32,
+    customer_demand_max: jax.numpy.int16,
+    max_start_window: jax.numpy.float32,
+    window_length: jax.numpy.float32,
+    early_coef_rand: Tuple[jax.numpy.float32, jax.numpy.float32],
+    late_coef_rand: Tuple[jax.numpy.float32, jax.numpy.float32],
+) -> Tuple[chex.Array, chex.Array, chex.Array, chex.Array, chex.Array, chex.Array]:
+
+    # Generate the node coordinates
+    coord_key, demand_key, window_key, earl_key, late_key = jax.random.split(key, 5)
+
+    node_coordinates = jax.random.uniform(
+        coord_key, (num_customers + 1, 2), minval=0, maxval=map_max
+    )
+
+    node_demands = jax.random.randint(
+        demand_key, (num_customers + 1,), minval=0, maxval=customer_demand_max
+    )
+    node_demands = node_demands.at[DEPOT_IDX].set(0)
+
+    # The total demands are controlled to less than the total capacity of all
+    # vehicles to ensure a feasible solution.
+    node_demands = jax.numpy.asarray(
+        node_demands * (total_capacity / jax.numpy.sum(node_demands)),
+        dtype=jax.numpy.int16,
+    )
+
+    window_start_times = jax.random.uniform(
+        window_key, (num_customers + 1,), minval=0, maxval=max_start_window
+    )
+    window_end_times = window_start_times + window_length
+
+    earl_coef = jax.random.uniform(
+        earl_key,
+        (num_customers + 1,),
+        minval=early_coef_rand[0],
+        maxval=early_coef_rand[1],
+    )
+    late_coef = jax.random.uniform(
+        late_key,
+        (num_customers + 1,),
+        minval=late_coef_rand[0],
+        maxval=late_coef_rand[1],
+    )
+
+    # No penalties for going to the depot
+    earl_coef = earl_coef.at[DEPOT_IDX].set(0.0)
+    late_coef = late_coef.at[DEPOT_IDX].set(0.0)
+
+    return (
+        node_coordinates,
+        node_demands,
+        window_start_times,
+        window_end_times,
+        earl_coef,
+        late_coef,
+    )
+
+
+def get_6_customer_init_settings(
```

Comment:
If this is for debugging/testing, should we move it to conftest?

Reply:
I updated the docstring. This is not just for testing, but serves as an simpler example to train on. When we update how actions can be masked the harder problems will become more tractable.